Data snapshot query execution

MishaDemianenko opened the pull request:
Introduce possibility to execute query based on data snapshot.
Any data modification will mark corresponding pages with a version that
will be equal to transaction id in which that particular change was introduced.
Also allow count store to track transactions in which changes where introduced.

To be able to guarantee that query result is based on a data that was
present on query execution start and was not modified while it was running,
the engine will verify that all paged that were accessed during query
execution have the version that is less or equal that was last closed
on a moment when query started.

Support for snapshot query execution in CC and HA

Snapshot query execution is enabled by 'unsupported.dbms.query.snapshot'
setting and number of retries that engine will do while trying to get
stable snapshot is controlled by 'unsupported.dbms.query.snapshot.retries'.

MishaDemianenko committed:
Data snapshot query execution

Introduce possibility to execute query based on data snapshot.
Any data modification will mark corresponding pages with a version that
will be equal to transaction id in which that particular change was introduced.
Also allow count store to track transactions in which changes where introduced.

To be able to guarantee that query result is based on a data that was
present on query execution start and was not modified while it was running,
the engine will verify that all paged that were accessed during query
execution have the version that is less or equal that was last closed
on a moment when query started.

Support for snapshot query execution in CC and HA

Snapshot query execution is enabled by 'unsupported.dbms.query.snapshot'
setting and number of retries that engine will do while trying to get
stable snapshot is controlled by 'unsupported.dbms.query.snapshot.retries'.

MishaDemianenko committed:
Do not clear context when statement is closed

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPage.java:
```diff
@@ -47,6 +50,10 @@
 
+    // max transaction id that updated this page
+    @SuppressWarnings( "unused" )
+    private volatile long lastModifiedTxId = UNBOUND_LAST_MODIFIED_TX_ID;
```
Since you're changing this field in some cases with `Unsafe`, could you take that all the way and always change/read it with `Unsafe` and remove `volatile`?

chrisvest replied:
I like keeping the `volatile` modifier as a reminder that this is the mode of access used on this field.

MishaDemianenko replied:
i would like (and actually did that), to keep it as volatile for exactly the same reason :)

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPage.java:
```diff
@@ -84,6 +91,32 @@ public long address()
+        return lastModifiedTxId;
+    }
+
+    public long resetLastModifiedTxId()
```
Would be nice if the name mentioned that it returned the txId, like `getAndResetLastModifiedTxId`

chrisvest replied:
And if a delegate for `getAndSetLong` is added to `UnsafeUtil`, then this method could be implemented in terms of it.

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCache.java:
```diff
@@ -196,6 +197,7 @@
     // 'true' (the default) if we should print any exceptions we get when unmapping a file.
     private boolean printExceptionsOnClose;
     private PageCursorTracerSupplier pageCursorTracerSupplier;
+    private VersionContextSupplier versionContextSupplier;
```
Can be final?

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -65,6 +66,7 @@
     private long pointer;
     private int pageSize;
     private int filePageSize;
+    protected VersionContext versionContext;
```
Can be final?

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPagedFile.java:
```diff
@@ -609,6 +610,25 @@ MuninnPage evictPage( long filePageId )
         return (MuninnPage) element;
     }
 
+    void setHighestEvictedTransactionId( long modifiedTransactionId )
```
Looks like a very similar method as `MuninnPage#setLastModifiedTxId`, could be made a utility with nice javadoc?

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPagedFile.java:
```diff
@@ -74,6 +75,10 @@
 
+    // max modifier transaction id among evicted pages for this file
+    private static final long evictedTransactionIdOffset = UnsafeUtil.getFieldOffset( MuninnPagedFile.class, "highestEvictedTransactionId" );
+    private volatile long highestEvictedTransactionId;
```
Also same note about `Unsafe` vs `volatile` as in `MuninnPage`

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnReadPageCursor.java:
```diff
@@ -60,10 +62,24 @@ public boolean next() throws IOException
+        {
+            return;
+        }
+        if ( page.getLastModifiedTxId() > versionContext.lastClosedTransactionId() ||
```
Would be very nice with human wording about what these conditions actually mean and how they play out.

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnReadPageCursor.java:
```diff
@@ -60,10 +62,24 @@ public boolean next() throws IOException
+            return;
+        }
+        if ( page.getLastModifiedTxId() > versionContext.lastClosedTransactionId() ||
+             pagedFile.getHighestEvictedTransactionId() > versionContext.lastClosedTransactionId() )
```
Also, read `versionContext.lastClosedTransactionId()` once instead of once per condition?

chrisvest replied:
If `MuninnPageCursor.pageFault` were to set the `lastModifiedTxId` to that of `pagedFile.getHighestEvictedTransactionId`, then we could omit the "highest evicted" check here, could we not? If so, then that would be one less volatile read.

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/context/EmptyVersionContext.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    public void initWrite( long committingTransactionId )
+    {
+
```
OCD newline

tinwelint commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/context/EmptyVersionContext.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    public void markAsDirty()
+    {
+
```
OCD newline

tinwelint commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java:
```diff
@@ -349,4 +565,82 @@ public void mustThrowIfMappingFileWouldOverflowReferenceCount() throws Exception
+        @Override
+        public void clear()
+        {
+
```
OCD newline

chrisvest commented on community/cypher/cypher/src/test/java/org/neo4j/cypher/internal/javacompat/EagerResultITTest.java:
```diff
@@ -0,0 +1,345 @@
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class EagerResultITTest
```
We have both IT and Test in the name. Probably one will be enough?

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageEvictionCallback.java:
```diff
@@ -38,5 +38,6 @@ public void onEvict( long filePageId, Page page )
         assert removed == page :
                 "Removed unexpected page when cleaning up translation table for filePageId " + filePageId + ". " +
                         "Evicted " + page + " but removed " + removed + " from the translation table.";
+        file.setHighestEvictedTransactionId( removed.resetLastModifiedTxId() );
```
I think we should do this before we `evictPage`, so there is no gap where "highest evicted" is lower than it should be.

chrisvest replied:
Oh, wait. We need the removed page object. In that case, I think this logic should be pushed down into `file.evictPage` so that we can capture this value before we remove the page from the translation table.

chrisvest replied:
I think this is important because while we are evicting the page, it is exclusively locked, so no one else can touch it, which in turn means that the "last modified tx id" will not change while we are evicting. The exclusive lock in turn means that as long as it remains in the translation table, no thread will initiate a page fault of that translation table entry, and therefor no thread will need the "last modified tx id" stored in the page object. So, by moving the tx id from the page object to the paged file, just before we remove the translation table entry, will ensure that no thread will miss out on any latest tx id.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -285,41 +287,20 @@ boolean applicable( DiagnosticsPhase phase )
+            AutoIndexing autoIndexing, PageCache pageCache, ConstraintSemantics constraintSemantics, Monitors monitors,
+            Tracers tracers, Procedures procedures, IOLimiter ioLimiter, AvailabilityGuard availabilityGuard,
+            SystemNanoClock clock, AccessCapability accessCapability, StoreCopyCheckPointMutex storeCopyCheckPointMutex,
+            IdController idController, VersionContextSupplier versionContextSupplier )
```
It becomes harder to see what's added or removed, when long parameter lists are collapsed like this.

MishaDemianenko replied:
well, the problem with that formatting is when you do not really introduce new parameter manually it's reformatted, but will return it to its previous form here since 3.2 and things...

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/ConfiguringPageCacheFactory.java:
```diff
@@ -54,12 +56,15 @@
      * @param pageCacheTracer global page cache tracer
      * @param pageCursorTracerSupplier supplier of thread local (transaction local) page cursor tracer that will provide
-     * thread local page cache statistics
+* thread local page cache statistics
```
Accidental removal of indentation.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsTracker.java:
```diff
@@ -84,46 +85,16 @@
-        }, new RotationTimerFactory( clock,
-                config.get( counts_store_rotation_timeout ).toMillis() ), 16, 16, HEADER_FIELDS );
+        super( fs, pages, baseFile, new CountsTrackerRotationMonitor( logProvider ), new RotationTimerFactory( clock,
+                config.get( counts_store_rotation_timeout ).toMillis() ), versionContextSupplier,16, 16, HEADER_FIELDS );
```
Missing a space in the parameter list, but otherwise 👍

chrisvest reviewed (changes requested):
Looks good, except I'd like you take a look at my suggestions around eviction and faulting.

chrisvest commented:
A note on forward merging: with the PageList being off-heap, we can choose to only spend the space on the version field, if versioning is enabled. Alternatively we could have a separate VersionList, which is only initialised when needed. The memory usage is slightly changed either way, however, which might be a problem for some tests in 3.4.

MishaDemianenko committed:
New year licenses.

MishaDemianenko committed:
Allow tracking version context on transaction that executed by multiple
threads (transactions that goes over rest end point).
Re-check version context when cursor repositioned to its current page again.

Check version context state after eager result visiting during serialization
since both bolt and rest writers will try to load all properties of
nodes and relationships from database, if those are part of result,
to send those to client as well. And since those executed outside of query
we need to expand version check to verify that after streaming result
we still based on valid snapshot.
If dirty context will be noticed exception will be thrown since we can't
restart that query - its too late and client already get partial results for it.

MishaDemianenko committed:
Formatting and naming comments.

MishaDemianenko committed:
Update new fields access to use unsafe only, tests for newly
introduced helper methods

chrisvest commented on community/unsafe/src/main/java/org/neo4j/unsafe/impl/internal/dragons/UnsafeUtil.java:
```diff
@@ -307,6 +307,32 @@ public static Object getAndSetObject( Object obj, long offset, Object newValue )
+        long currentValue;
+        do
+        {
+            currentValue = UnsafeUtil.getLong( object, fieldOffset );
```
I think this needs to be a `getLongVolatile`.

MishaDemianenko replied:
yes, thx that's a mistake it should be

MishaDemianenko committed:
Use correct volatile long get instead of simple long get

MishaDemianenko committed:
Update tracking of highest page file version.

Make timeout for CC test a bit longer.

MishaDemianenko commented:
@chrisvest @tinwelint now everything should be good and ready for merge.
Note: will be null merged into 3.3 and 3.4. Separate PR will be created since some additional testing (for 3.3) and rewrite for 3.4 required.

MishaDemianenko committed:
Give a bit more time for clustering test
